### PR TITLE
Fix readme, add double-checked locking on SpanCollector

### PR DESCRIPTION
**Pull request by klettier**
As **SpanCollector.GetInstance** can be called by multiple threads at the same time I think it's more safe to add double-checked locking in singleton creation expression.

### CODE REVIEW

**Comment by klettier** [comment]
I want to discuss about some modifications I would like to add, could you please enable Issues on your repository to allow us to share ?

Thx

**Comment by cabbott** [comment]
@klettier issues enabled!

**Comment by jcarres-mdsol** [comment]
@bvillanueva-mdsol  @lschreck-mdsol  @kenyamat  お願いします 🙏

File: src/Medidata.ZipkinTracer.Core/SpanCollector.cs [hunk 1]
```diff
@@ -13,13 +13,20 @@ public class SpanCollector
...
 
         public static SpanCollector GetInstance(Uri uri, uint maxProcessorBatchSize, ILog logger)
         {
             if (instance == null)
             {
-                instance = new SpanCollector(uri, maxProcessorBatchSize, logger);
-                instance.Start();
+                lock (syncObj)
+                {
+                    if (instance == null)
```
Comment (jcarres-mdsol) [inline src/Medidata.ZipkinTracer.Core/SpanCollector.cs]: already inside a `  if (instance == null)` block

Comment (klettier) [inline src/Medidata.ZipkinTracer.Core/SpanCollector.cs]: How do you guarantee that two threads haven't already go through this condition before the instance was effectively created?

**Comment by bvillanueva-mdsol** [comment]
@klettier I think this looks good. I thought also the same with what @jcarres-mdsol commented, but I realized it is a good a double check thread-safe style for singleton. Talked with @lschreck-mdsol before about avoiding the use of ```lock()``` and use lazy instantiation instead. That would be preferable but the parameters of ```SpanCollector``` is not from config or any static property but supplied using the ```ZipkinClient``` constructor parameters. @lschreck-mdsol Do you have any idea on another way? 

@klettier Please update branch from main develop.

Thanks!

**Comment by klettier** [comment]
Hi @bvillanueva-mdsol,

For me, is not the responsability of SpanCollector to maintain a singleton instance of itself.
The caller (ZipkinClient) has to have that responsiility because he makes the choice of passing the constructor instance or to give the default one.
Don't be afraid by the lock because it will occur only one time so it's not going to have performance impact the important thing here is to protect your instance from being call by multiple threads at the same time.

File: src/Medidata.ZipkinTracer.Core/Helpers/SyncHelper.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Collections.Generic;
```
Comment (bvillanueva-mdsol) [inline src/Medidata.ZipkinTracer.Core/Helpers/SyncHelper.cs]: minor comment: @klettier Please delete unused usings here. Thanks!

File: src/Medidata.ZipkinTracer.Core/SpanCollector.cs [hunk 3]
```diff
@@ -41,12 +33,12 @@ public virtual void Collect(Span span)
 
         public virtual void Start()
         {
-            spanProcessor.Start();
+            SyncHelper.ExecuteSafely(syncObj, () => !IsStarted, () => { spanProcessor.Start(); IsStarted = true; });
```
Comment (bvillanueva-mdsol) [inline src/Medidata.ZipkinTracer.Core/SpanCollector.cs]: @klettier Can we bring ExecuteSafely() method at the SpanProcessorTaskFactory level? I think it has the similar logic for IsStarted there. it checks the spanProcessorTaskInstance if it is already instantiated.

File: src/Medidata.ZipkinTracer.Core/Helpers/SyncHelper.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Medidata.ZipkinTracer.Core.Helpers
+{
+   public static class SyncHelper
+    {
+        public static void ExecuteSafely(object sync, Func<bool> canExecute, Action actiontoExecuteSafely)
```
Comment (bvillanueva-mdsol) [inline src/Medidata.ZipkinTracer.Core/Helpers/SyncHelper.cs]: :+1:

File: src/Medidata.ZipkinTracer.Core/ZipkinClient.cs [hunk 5]
```diff
@@ -35,10 +50,12 @@ public ZipkinClient(IZipkinConfig zipkinConfig, IOwinContext context, SpanCollec
 
             try
             {
-                spanCollector = collector ?? SpanCollector.GetInstance(
+                spanCollector = collector ?? GetInstance(
```
Comment (bvillanueva-mdsol) [inline src/Medidata.ZipkinTracer.Core/ZipkinClient.cs]: @klettier There was a miss in unit test scenario for null collector. Can you please add one scenario to cover the GetInstance part? Thanks!

**Comment by bvillanueva-mdsol** [comment]
@klettier I agree that ZipkinClient should have the responsibility to take are of the Singleton instance. I like the the helper method you made. Please check out my comments on the code. Thanks!

**Comment by klettier** [comment]
@bvillanueva-mdsol Thx for your comments
I have more ideas on what we can do next so I will work on them and submit some issues soon ;)

**Comment by bvillanueva-mdsol** [comment]
@klettier This looks good. With this changes, we are planning to run a performance test soon just for verification. Performance test are executed to make sure we don't interfere with main operation. I'll ask for allocation of time for our next sprint on the performance tests. Will keep you posted on schedule if we can secure it as well as the results. Thank you so much!

**Comment by bvillanueva-mdsol** [comment]
@klettier Unfortunately we have things line up for this sprint and might not be able to touch on the performance test for the next two weeks. I'll get back to you after then. For the mean time, can please merge latest changes from develop in this branch. Thanks!

**Comment by thoshikawa-mdsol** [comment]
@klettier @mdsol/team-51 

I've run performance test on Ampridatvir with this change.
All the counters look good, so there is no performance regressions by this change.

image

image

**Michine resource comparison**
image

Full run to run comparison result is available [here](https://drive.google.com/open?id=0B7DORCTPoHT5Q3ZzNDRQbTRVWHM)

**Comment by bvillanueva-mdsol** [comment]
Looks good @thoshikawa-mdsol . Thanks for sharing the performance test data.

**Comment by klettier** [comment]
Looks good, thanks @thoshikawa-mdsol

**Comment by bvillanueva-mdsol** [comment]
Merging. Thanks @klettier for this PR. :+1:

**Pull request merged**